src/renderers/shaders/ShaderChunk/lights_phong_fragment.glsl

```diff
@@ -78,32 +51,37 @@ vec3 totalSpecularLight = vec3( 0.0 );
 
 	for( int i = 0; i < MAX_DIR_LIGHTS; i ++ ) {
 
-		vec3 lightColor = directionalLightColor[ i ];
+		IncidentLight incidentLight;
+		getDirLightDirect( directionalLights[ i ], incidentLight );
 
-		vec3 lightDir = directionalLightDirection[ i ];
+		totalDirectReflectedDiffuse +=
+			BRDF_Lambert( incidentLight, normal, diffuse );
 
-		// diffuse
+		totalDirectReflectedSpecular +=
+			BRDF_BlinnPhong( incidentLight, normal, viewDir, specular, shininess );
 
-		float cosineTerm = saturate( dot( normal, lightDir ) );
+	}
 
-		totalDiffuseLight += lightColor * cosineTerm;
+#endif
+
+#if MAX_HEMI_LIGHTS > 0
 
-		// specular
+	for( int i = 0; i < MAX_HEMI_LIGHTS; i ++ ) {
 
-		vec3 brdf = BRDF_BlinnPhong( specular, shininess, normal, lightDir, viewDir );
+		IncidentLight incidentLight;
+		getHemisphereLightIndirect( hemisphereLights[ i ], normal, incidentLight );
 
-		totalSpecularLight += brdf * specularStrength * lightColor * cosineTerm;
+		totalIndirectReflectedDiffuse +=
+			BRDF_Lambert( incidentLight, normal, diffuse );
 
 	}
 
 #endif
 
-#ifdef METAL
-
-	outgoingLight += diffuseColor.rgb * ( totalDiffuseLight + totalAmbientLight ) * specular + totalSpecularLight + totalEmissiveLight;
 
-#else
-
-	outgoingLight += diffuseColor.rgb * ( totalDiffuseLight + totalAmbientLight ) + totalSpecularLight + totalEmissiveLight;
-
-#endif
+outgoingLight +=
+	totalDirectReflectedDiffuse +
+	totalDirectReflectedSpecular +
+	totalIndirectReflectedDiffuse +
+	totalIndirectReflectedSpecular +
+	totalEmissiveLight;
```

Comment:
Is this really such a good idea, instead of accumulating `outgoingLight` right away?

A smart GLSL compiler will optimize it right. Less smart (true ES, on-target) compilers may take the code more literally. In this case, spilling registers to slower memory, limiting the number of parallel threads, or both.